flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java

```diff
@@ -351,38 +347,68 @@ ByteBuf write(ByteBufAllocator allocator) throws IOException {
 
 				CompositeByteBuf composityBuf = allocator.compositeDirectBuffer();
 				composityBuf.addComponent(headerBuf);
-				composityBuf.addComponent(buffer);
+				composityBuf.addComponent(buffer.asByteBuf());
 				// update writer index since we have data written to the components:
-				composityBuf.writerIndex(headerBuf.writerIndex() + buffer.writerIndex());
+				composityBuf.writerIndex(headerBuf.writerIndex() + buffer.asByteBuf().writerIndex());
 				return composityBuf;
 			}
 			catch (Throwable t) {
 				if (headerBuf != null) {
 					headerBuf.release();
 				}
-				buffer.release();
+				buffer.recycleBuffer();
 
 				ExceptionUtils.rethrowIOException(t);
 				return null; // silence the compiler
 			}
 		}
 
-		static BufferResponse readFrom(ByteBuf buffer) {
-			InputChannelID receiverId = InputChannelID.fromByteBuf(buffer);
-			int sequenceNumber = buffer.readInt();
-			int backlog = buffer.readInt();
-			boolean isBuffer = buffer.readBoolean();
-			boolean isCompressed = buffer.readBoolean();
-			int size = buffer.readInt();
+		/**
+		 * Parses the message header part and composes a new BufferResponse with an empty data buffer. The
+		 * data buffer will be filled in later. This method is used in credit-based network stack.
+		 *
+		 * @param messageHeader the serialized message header.
+		 * @param bufferAllocator the allocator for network buffer.
+		 * @return a BufferResponse object with the header parsed and the data buffer to fill in later. The
+		 *			data buffer will be null if the target channel has been released or the buffer size is 0.
+		 */
+		static BufferResponse readFrom(ByteBuf messageHeader, NetworkBufferAllocator bufferAllocator) {
+			InputChannelID receiverId = InputChannelID.fromByteBuf(messageHeader);
+			int sequenceNumber = messageHeader.readInt();
+			int backlog = messageHeader.readInt();
+			boolean isBuffer = messageHeader.readBoolean();
+			boolean isCompressed = messageHeader.readBoolean();
+			int size = messageHeader.readInt();
+
+			Buffer dataBuffer = null;
+			if (size != 0) {
+				if (isBuffer) {
+					dataBuffer = bufferAllocator.allocatePooledNetworkBuffer(receiverId);
```

Comment:
In the past we requested the flink buffer from channel after all the whole message bytes are accumulated. Now we request the flink buffer beforehand. That means this buffer would be idle some time if the following data bytes are not received from network immediately. If it is a floating buffer, it might hurt throughput to some extent.